### PR TITLE
Update properties before upgrading custom elements (Fixes #453)

### LINKED ISSUE #453: Custom elements should be connected after being updated
There's currently a difference between `TemplateResult`'s processed via `render` and those handled by `NodePart` in a "nested" `TemplateResult`. The former updates prior to being attached to the DOM, but the latter does not. 

This can cause an issue if a custom element processes state in the `connectedCallback`. Elements made with `Polymer` and `LitElement` (prior to version `0.6`) do this, and when used in a nested `TemplateResult` these elements render with default state and then update with the state set by `lit-html`. This is inconsistent with non-nested `TemplateResult`'s and less efficient. 

The fix is simple, just update the instance before https://github.com/Polymer/lit-html/blob/master/src/lib/parts.ts#L267.

### PR DESCRIPTION
**Pull request by ruphin**
This will set the values in templates before elements are upgraded.

### CODE REVIEW

File: src/lib/parts.ts [hunk 1]
```diff
@@ -249,16 +249,16 @@ export class NodePart implements Part {
...
-    let instance: TemplateInstance;
     if (this.value && this.value.template === template) {
-      instance = this.value;
+      this.value.update(value.values);
     } else {
       // Make sure we propagate the template processor from the TemplateResult
       // so that we use it's syntax extension, etc. The template factory comes
       // from the render function so that it can control caching.
-      instance =
+      const instance =
```
Comment (justinfagnani) [inline src/lib/parts.ts]: This should come after elements are upgraded, right before `this._commitNode(fragment);`. That should make the tests pass.

Comment (ruphin) [reply src/lib/parts.ts]: A render test in Firefox is still breaking. I'm looking into it

Comment (justinfagnani) [reply src/lib/parts.ts]: We don't run Sauce tests on PRs, so FF is the only configuration that tests the polyfills right now. It's probably also failing on Edge on Sauce.

**Comment by ruphin** [comment]
I found the problem; we were not upgrading fragments in parts that were not connected, but when we update parts before connecting nodes, this causes nested elements to not upgrade.

**Comment by ruphin** [comment]
Looking into this more closely I feel like we can also change `TemplateInstance._clone()` to return a DocumentFragment that has already been imported. I think every consumer of `_clone()` more or less immediately imports the fragment and/or upgrades the elements now, so we get some nice deduplication by incorporating this into `_clone()`

**Comment by ruphin** [comment]
Fixes #453

@justinfagnani ptal :)

**Comment by justinfagnani** [comment]
> Looking into this more closely I feel like we can also change TemplateInstance._clone() to return a DocumentFragment that has already been imported.

That's a good idea :)

**Comment by ruphin** [comment]
I'll make a new PR that implements the suggested change

**Pull request merged**